smarts/core/smarts.py

```diff
@@ -555,7 +555,7 @@ def teardown(self):
 
     def destroy(self):
         self.teardown()
-
+        self._vehicle_index = None
```

Comment:
This line is needed to make sure that `SMARTS` does not try to call the vehicle index cleanup code again.

Reply:
Is the line `self._vehicle_index=None` included here so that when `destroy()` is called again, the `self._vehicle_index.teardown()` inside `teardown()` will not be called again?

Why and how is the `destroy()` function being called multiple times within a code?

Reply:
> Is the line self._vehicle_index=None included here so that when destroy() is called again, the self._vehicle_index.teardown() inside teardown() will not be called again?

Yes, although, I think there is the potential alternative that instead the `_vehicle_index` could be set to `None` in the `teardown()` method and then no special action is required in the `destroy()` method.

> Why and how is the destroy() function being called multiple times within a code?

The problem case is if `destroy()` is called and then the `SMARTS` instance reaches the end of the program and is cleaned up by uncontrolled garbage collection where some modules might not be available by the time the instance `__del__()` is called.

However, I am not sure if we should remove the `destroy()` call from `__del__()` because putting `destroy()` in `__del__()` allows `del` to also properly clean up a `SMARTS` instance, and provides a shorthand (using the exception catch) for figuring out if `SMARTS` has already been properly cleaned up.

Reply:
> The problem case is if destroy() is called and then the SMARTS instance reaches the end of the program and is cleaned up by uncontrolled garbage collection where some modules might not be available by the time the instance __del__() is called.

Reply:
> This line is needed to make sure that SMARTS does not try to call the vehicle index cleanup code again.

Sorry, I think my original comment was misleading, even though this is the effect of the change. The current version of `destroy()` and `teardown()` can be called any number of times without problem. The crashing issue with `destroy()` or `teardown()` only occurs at the end of program garbage collection. I can note the following cases:

1) No Error: `del smarts` is called which calls `__del__()` removing the SMARTS object before it reaches end of program.
2) Fixed error: `destroy()` is called which then will cause an error at end of program because the attribute `_vehicle_index` was not set to `None` and it will access modules that may already be deleted.
3) Error: Neither are called and `destroy()` throws an error  caused by end of program garbage collection which is caught by `__del__()` and then asks the user to call `destroy()` explicitly.

To do exactly what you are suggesting I would need to repurpose `__del__()` to only check if `destroy()` was previously called before `__del__()` and then disallow multiple calls to `destroy()`. This might be reasonable, but I am on the fence about it because it removes the `del smarts_obj` syntax.

Reply:
> However, I am not sure if we should remove the `destroy()` call from `__del__()` because putting `destroy()` in `__del__()` allows `del` to also properly clean up a `SMARTS` instance, and provides a shorthand (using the exception catch) for figuring out if `SMARTS` has already been properly cleaned up.

What about the SMARTS object needs to be "properly cleaned up" (as opposed to just cleaned up with normal garbage collection)?  I.e., I'd like to understand why normal garbage collection is not enough for us (either in the normal shutdown case or in the exception case).

And whatever this is should be clearly documented in a comment in `__del__()` (and/or `destroy()`).

Reply:
I believe that the main sources of problem requiring cleanup are port, file, and subprocess related: TraCI and grpc. I believe the `SumoTrafficSimuation` and `AgentManager` are the core parts that would want guaranteed shutdown.

I think to get the exact specifics would need some additional information. As it stands, I am now leaning towards removing the cleanup from `__del__()` and requiring `destroy()` to be called.

Reply:
OK, yeah, I think you're right about Sumo Traffic and the Remote (zoo) Agents, both of which spawn sub-processes that need to be killed.  It might be worth adding a comment about that in either `__del__()` or `destroy()`.

I think we should make both `destroy()` and `__del__()`  as idempotent as possible.  I think your changes here (in particular, setting `self._vehicle_index` to `None`) achieve that, or at least improve on the situation, so I'm inclined to approve this "as is" (with the addition of the comment I requested).

The nice part of keeping the cleanup (call to `destroy()`) in `__del__()` too is that it allows for multiple SMARTS objects to be created (without having to require the user to explicitly call `destroy()` in between, which is an oddity in general) without leaking processes.  But I would support either approach if you guys feel strongly one way or another.